underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java

```diff
@@ -0,0 +1,406 @@
+
+  private static final boolean SSE_ENABLED =
+      Configuration.getBoolean(PropertyKey.UNDERFS_S3A_SERVER_SIDE_ENCRYPTION_ENABLED);
+
+  /**
+   * Only parts bigger than 5MB could be uploaded through multipart upload,
+   * except the last part.
+   */
+  private static final long UPLOAD_THRESHOLD = 5 * Constants.MB;
+
+  /** Bucket name of the Alluxio S3 bucket. */
+  private final String mBucketName;
+
+  /** The Amazon S3 client responsible to upload data to S3. */
+  private final AmazonS3 mClient;
+
+  /** Executing the upload tasks. */
+  private final ListeningExecutorService mExecutor;
+
+  /** Key of the file when it is uploaded to S3. */
+  private final String mKey;
+
+  /** Allowed size of the partition. */
+  private final long mPartitionSize;
+
+  /** Pre-allocated byte buffer for writing single characters. */
+  private final byte[] mSingleCharWrite = new byte[1];
+
+  /** Tags for the uploaded part, provided by S3 after uploading. */
+  private final List<PartETag> mTags = new ArrayList<>();
+
+  /** The retry policy of this multipart upload. */
+  private final RetryPolicy mRetryPolicy = new CountingRetry(5);
+
+  /** Flag to indicate this stream has been closed, to ensure close is only done once. */
+  private boolean mClosed = false;
+
+  /**
+   * The local file that will be uploaded when reaches the partition size
+   * or when flush() is called and this file is bigger than 5MB.
+   */
+  private File mFile;
+
+  /** Store the future of tags. */
+  private List<ListenableFuture<PartETag>> mFutureTags = new ArrayList<>();
+
+  /** The MD5 hash of the file. */
+  private MessageDigest mHash;
+
+  /** The output stream to a local file where the file will be buffered. */
+  private OutputStream mLocalOutputStream;
+
+  /** When the offset reaches the partition size, we upload the temp file. */
+  private long mOffset;
+
+  /**
+   * Give each upload request an unique and continuous id
+   * so that S3 knows the part sequence to concatenate the parts to a single object.
+   */
+  private AtomicInteger mPartNumber;
+
+  /** The upload id of this multipart upload. */
+  private String mUploadId;
+
+  /**
+   * Constructs a new stream for writing a file.
+   *
+   * @param bucketName the name of the bucket
+   * @param key the key of the file
+   * @param s3Client the Amazon S3 client to upload the file with
+   * @param executor a thread pool executor
+   */
+  public S3ALowLevelOutputStream(String bucketName, String key, AmazonS3 s3Client,
+      ExecutorService executor) {
+    Preconditions.checkArgument(bucketName != null && !bucketName.isEmpty(), "Bucket name must "
+        + "not be null or empty.");
+    mBucketName = bucketName;
+    mClient = s3Client;
+    mExecutor = MoreExecutors.listeningDecorator(executor);
+    try {
+      mHash = MessageDigest.getInstance("MD5");
+    } catch (NoSuchAlgorithmException e) {
+      LOG.warn("Algorithm not available for MD5 hash.", e);
```

Comment:
this should never happen, can throw RuntimeException

Reply:
Agree, this should never happen, but throw an exception in the constructor should be avoided.
Putting mHash initialization in other methods looks wired, I suggest remaining it here.